### PR TITLE
#19527 - This comes to solve the 10K items count limitation on ES.

### LINKED ISSUE #19527: Rely on Elasticsearch index.max_result_window size to dictate the max limit of query results
**Describe the bug**

Reproduced on demo.  Reported by customer: @https://dotcms.zendesk.com/agent/tickets/102760

When you run `$dotcontent.pull()`, if you attempt to display any content past the 10,000th result, no results are returned.

This is true even if you supply an offset and reasonable limit.  For example, the following will never return any results, even if you have more than 10,000 content items of the content type:

```
$dotcontent.pull( "+contentType:ctVarName", 9999, 10, "modDate desc" )
```

**To Reproduce**

1. Create a simple content type, with just a Title field.
    * User Searchable, Show in List, **_not_** Unique.
2. Create more than 10,000 items of the content type.  For example:
    * Create one content item
    * Export it
    * Open the CSV in a spreadsheet program
    * Copy the line 11,000 times.
    * Use a formula to set the Title so it increment for each row (e.g. `="Title "&(row(A2)-1)`)
    * Save the spreadsheet as CSV
    * Import the CSV
        * Note: Depending on where you're running, you may need to break the CSV into multiple files to be able to import it without the import timing out.
3. Working: Add a page that uses a content pull with a limit of 10 and an offset of 9990 to display all the results of the pull.  Example:
```
        <ul>
        #foreach( $item in $dotcontent.pull( "+contentType:SuperSimple", 9990, 10, "title asc" ) )
            <li>$item.title</li>
        #end
        </ul>
```
4. Not working: Change the offset from 9990 to 9991 and display the page again.
    * **Result:** No results are displayed.

**Expected behavior**

Contents past the 10,000th result should be returned.  A hard-coded maximum for the limit makes sense, but does not make sense for the offset.

### PR DESCRIPTION
**Pull request by fabrizzio-dotCMS**
Our index count methods are executed via  ES's `SearchRequest` which is pretty much limited by the 10K items hard limit restriction. I am replacing the uses of `SearchRequest` with `CountRequest` which operates without the  10K items restriction.

### CODE REVIEW

**Review by nollymar — COMMENTED** [review]
The change seems to be ok. I would remove the unused methods and add some ITs

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java [hunk 1]
```diff
@@ -1489,55 +1454,60 @@ protected void indexCount(final String query,
...
+       // we check the query to figure out which indexes to hit
+
+       final IndiciesInfo info;
+       try {
+           info = APILocator.getIndiciesAPI().loadIndicies();
+       } catch (DotDataException e) {
+           throw new DotRuntimeException(e);
+       }
+
+       final String indexToHit;
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java:1502]: I would change this to

return query.contains("+live:true") && !query.contains("+deleted:true")? info.getLive(): info.getWorking()

**Review by jdotcms — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java [hunk 2]
```diff
@@ -1631,7 +1629,40 @@ SearchHits cachedIndexSearch(final SearchRequest searchRequest) {
...
         
     }
-    
+
+    /**
+     * if enabled CountRequest are executed and then cached
+     * @param countRequest
+     * @return
+     */
+    Long cachedIndexCount(final CountRequest countRequest) {
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java:1638]: Do we need this if we are getting the correct total back from the SearchHits?

File: dotCMS/src/main/resources/dotmarketing-config.properties [hunk 3]
```diff
@@ -522,6 +522,14 @@ CACHE_INVALIDATION_TRANSPORT_CLASS=com.dotcms.cache.transport.HazelcastCacheTran
 ## Cache elasticsearch query results
 ES_CACHE_SEARCH_QUERIES=true
 
+# This Property can be set to a positive numeric value or a boolean value
+# if set to false ES will stop retrieving the hits Count. DO NOT SET IT TO FALSE.
+# IF set to true  ES will track all hits beyond the 10K items limit. This might imply an impact performance wise.
+# By default we're providing a max number.
+# If commented out ES will default to 10K again.
+# https://www.elastic.co/guide/en/elasticsearch/reference/7.10/search-your-data.html#track-total-hits
+ES_TRACK_TOTAL_HITS=1000000
```
Comment (wezell) [inline dotCMS/src/main/resources/dotmarketing-config.properties:531]: Our default should be 1,000,000 and this should not be set by default - we should not need to add this property to the `dotmarketing-config.properties` file.

**Review by wezell — CHANGES_REQUESTED** [review]
Let's talk about these changes

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java [hunk 4]
```diff
@@ -1430,56 +1433,18 @@ protected Identifier getRelatedIdentifier(Contentlet contentlet, String relation
...
-            return 0;
-        }
-
-        SearchRequest searchRequest = getCountSearchRequest(queryStringQuery);
-        searchRequest.indices(query.contains("+live:true") && !query.contains("+deleted:true")?
-                info.getLive(): info.getWorking());
-
-        final SearchHits hits = cachedIndexSearch(searchRequest);
-        return hits.getTotalHits().value;
+       return indexCount(query);
```
[the remark this answers is not in the document]
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java]: nope.

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java (hunk 2 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java:1599]: This one is always accurate. and does not depend on the `track_total_hits`

**Review by nollymar — APPROVED** [review]

**Review by erickgonzalez — APPROVED** [review]

**Pull request merged**